File: AlgorithmicToolbox/FibonacciNumberAgain.py
# Problem Description: Compute the n-th Fibonacci number modulo m.

import doctest


def fibonaccimodulo(n, m):
    """Returns nth fibonacci number modulo m.

    Examples:
    >>> fibonaccimodulo(1, 239)
    1
    >>> fibonaccimodulo(115, 1000)
    885
    >>> fibonaccimodulo(2816213588, 30524)
    10249
    """

    f = [0, 1]
    reps = n+1
    for i in range(2, min(m**2, n+1)):
        curr = (f[i-1] + f[i-2]) % m
        if(i == n):
            return curr
        if(curr == 0 and f[i-1] == 1):
            reps = len(f)
            break
        f.append(curr)
    return f[n % reps]


if __name__ == "__main__":
    doctest.testmod()
    n, m = map(int, input().split())
    print(fibonaccimodulo(n, m))

File: AlgorithmicToolbox/FibonacciNumberAgain.cpp
// Problem Description: Compute the n-th Fibonacci number modulo m.

#include <iostream>
#include <vector>
using namespace std;

long long fibonaccimodulo(long long n, long long m)
{
    long long curr;
    long reps = n + 1;
    vector<long long> f = {0, 1};
    for (int i = 2; i < m * m && i <= n; i++)
    {
        curr = (f[i - 1] + f[i - 2]) % m;
        if (i == n)
        {
            return curr;
        }
        if (curr == 0 && f[i - 1] == 1)
        {
            reps = f.size();
            break;
        }
        f.push_back(curr);
    }
    return (f[n % reps]);
}

int main()
{
    long long n, m;
    cin >> n >> m;
    cout << fibonaccimodulo(n, m) << endl;
    return 0;
}
